### PR TITLE
Add event for player editing sign

### LINKED ISSUE #9293: Player Edit Sign Event
### Is your feature request related to a problem?

Cancelling PlayerInteractEvent might be limiting because there's a somewhat complicated logic for determining which side of sign the player clicked. (See nms `SignBlockEntity#isFacingFrontText`)

### Describe the solution you'd like.

A event that triggers when a player tries to edit a sign. With a way to get which side of the sign is being clicked. Cancel the event to do nothing when clicked (don't open the sign edit screen)

Edit: After seeing @lynxplay's idea, this event maybe should also trigger after placing the sign. Cancelling the event would not open the sign edit screen when placed.

### Describe alternatives you've considered.

I mean, I can copy the nms logic in my plugin code but that's kinda bad

### Other

_No response_

### PR DESCRIPTION
**Pull request by by77er**
resolves https://github.com/PaperMC/Paper/issues/9293

### CODE REVIEW

File: patches/api/0419-Add-event-for-player-editing-sign.patch [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
++import org.bukkit.event.Cancellable;
++import org.bukkit.event.HandlerList;
++import org.bukkit.event.player.PlayerEvent;
++import org.jetbrains.annotations.NotNull;
++
++/**
++ * Called when a player edits a sign
++ */
++public class PlayerEditSignEvent extends PlayerEvent implements Cancellable {
++    private static final HandlerList handlers = new HandlerList();
```
Comment (Machine-Maker) [inline patches/api/0419-Add-event-for-player-editing-sign.patch]: `HANDLER_LIST`

File: patches/api/0419-Add-event-for-player-editing-sign.patch [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
++        this.side = side;
++    }
++
++    /**
++     * Get the sign that was clicked
++     *
++     * @return {@link Sign} that was clicked
++     */
++    @NotNull
++    public Sign getSign() { return sign; }
```
Comment (Machine-Maker) [inline patches/api/0419-Add-event-for-player-editing-sign.patch]: No need to make these methods in one line

File: patches/api/0419-Add-event-for-player-editing-sign.patch [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
++    @NotNull
++    public Side getSide() { return side; }
++
++    public boolean isCancelled() {
++        return cancel;
++    }
++
++    public void setCancelled(boolean cancel) {
++        this.cancel = cancel;
++    }
```
Comment (Machine-Maker) [inline patches/api/0419-Add-event-for-player-editing-sign.patch]: Override annotations

File: patches/server/0977-Add-event-for-player-editing-sign.patch [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
++        Item potentialApplicator = itemStack.getItem(); // Paper: renamed signBlockEntity -> potentialApplicator due to decompile error
+         SignApplicator var10000;
+-        if (signBlockEntity instanceof SignApplicator signApplicator) {
++        if (potentialApplicator instanceof SignApplicator signApplicator) { // Paper: renamed signBlockEntity -> potentialApplicator due to decompile error
+             var10000 = signApplicator;
+         } else {
+             var10000 = null;
+@@ -86,8 +86,8 @@ public abstract class SignBlock extends BaseEntityBlock implements SimpleWaterlo
+ 
+         SignApplicator signApplicator2 = var10000;
```
Comment (Machine-Maker) [inline patches/server/0977-Add-event-for-player-editing-sign.patch]: This whole block is super confusing, idk what happened to the decompiler here. Is it worth just deleting it and rewriting it to make actual sense? Looking for input here, not saying it should be done.

File: patches/server/0977-Add-event-for-player-editing-sign.patch [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+                 SignText signText = signBlockEntity.getText(bl2);
+@@ -153,8 +153,17 @@ public abstract class SignBlock extends BaseEntityBlock implements SimpleWaterlo
+     }
+ 
+     public void openTextEdit(Player player, SignBlockEntity blockEntity, boolean front) {
+-        blockEntity.setAllowedPlayerEditor(player.getUUID());
+-        player.openTextEdit(blockEntity, front);
++        // Paper start
++        var bukkitPlayer = (org.bukkit.entity.Player) player.getBukkitEntity();
++        var event = new com.destroystokyo.paper.event.player.PlayerEditSignEvent(
```
Comment (Machine-Maker) [inline patches/server/0977-Add-event-for-player-editing-sign.patch]: We aren't big fans of `var` even if it saves a fully qualified name.

File: patches/server/0977-Add-event-for-player-editing-sign.patch [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
++            bukkitPlayer,
++            new org.bukkit.craftbukkit.block.CraftSign<>(bukkitPlayer.getWorld(), blockEntity),
```
Comment (Machine-Maker) [inline patches/server/0977-Add-event-for-player-editing-sign.patch]: This isn't correct, it could be a hanging sign. CraftSign doesn't cover that. You should instead use one of the methods on `CraftBlockStates` to get the blockstate instance.

**Review by Machine-Maker — DISMISSED** [review]

File: patches/server/0977-Add-event-for-player-editing-sign.patch (hunk 6 above)
Comment (by77er) [reply patches/server/0977-Add-event-for-player-editing-sign.patch]: In working on this I realized it might be better to return a `Block` in place of the `Sign`, since the event handler probably needs more context anyway. Thoughts?

Comment (Machine-Maker) [reply patches/server/0977-Add-event-for-player-editing-sign.patch]: Sign is fine as HangingSign extends it, but you need to create a CraftHangingSign if its a hanging sign instead of CraftSign for a regular sign.

File: patches/api/0419-Add-event-for-player-editing-sign.patch (hunk 1 above)
Comment (by77er) [reply patches/api/0419-Add-event-for-player-editing-sign.patch]: Done!

File: patches/api/0419-Add-event-for-player-editing-sign.patch (hunk 2 above)
Comment (by77er) [reply patches/api/0419-Add-event-for-player-editing-sign.patch]: Expanded.

File: patches/api/0419-Add-event-for-player-editing-sign.patch (hunk 3 above)
Comment (by77er) [reply patches/api/0419-Add-event-for-player-editing-sign.patch]: Done!

File: patches/server/0977-Add-event-for-player-editing-sign.patch (hunk 5 above)
Comment (by77er) [reply patches/server/0977-Add-event-for-player-editing-sign.patch]: Replaced with fully qualified names.

File: patches/api/0419-Add-event-for-player-editing-sign.patch (hunk 3 above)
Comment (by77er) [reply patches/api/0419-Add-event-for-player-editing-sign.patch]: Added.

File: patches/server/0977-Add-event-for-player-editing-sign.patch (hunk 6 above)
Comment (by77er) [reply patches/server/0977-Add-event-for-player-editing-sign.patch]: Alright, let me know if it looks good now.

**Comment by molor** [comment]
Also wrong (outdated) package name

File: patches/server/0977-Add-event-for-player-editing-sign.patch (hunk 4 above)
Comment (lynxplay) [reply patches/server/0977-Add-event-for-player-editing-sign.patch]: Yea I honestly wouldn't mind just yanking it.
I mean, its gonna get diffed on every update that touches it anyway with the amount of decompiler fixes it requires.

Comment (lynxplay) [reply patches/server/0977-Add-event-for-player-editing-sign.patch]: Both 
```patch
-        Item signBlockEntity = itemStack.getItem();
-        SignApplicator var10000;
-        if (signBlockEntity instanceof SignApplicator signApplicator) {
-            var10000 = signApplicator;
-        } else {
-            var10000 = null;
-        }
-
-        SignApplicator signApplicator2 = var10000;
+        SignApplicator signApplicator2 = item instanceof SignApplicator signApplicator ? signApplicator : null; // Paper - decompile fixes
```

and 
```patch
-        BlockEntity bl2 = world.getBlockEntity(pos);
-        if (bl2 instanceof SignBlockEntity signBlockEntity) {
+        if (world.getBlockEntity(pos) instanceof SignBlockEntity signBlockEntity) { // Paper - decompile fixes
```
would probably do it.

File: patches/api/0419-Add-event-for-player-editing-sign.patch [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
++import org.bukkit.event.HandlerList;
++import org.bukkit.event.player.PlayerEvent;
++import org.jetbrains.annotations.NotNull;
++
++/**
++ * Called when a player edits a sign
++ * <p>
++ * Cancelling this event stops the sign editing menu from appearing.
++ */
++public class PlayerEditSignEvent extends PlayerEvent implements Cancellable {
```
Comment (lynxplay) [inline patches/api/0419-Add-event-for-player-editing-sign.patch]: I wonder if this name properly communicates the fact that this event isn't really called when the edit happens but when the player opens the editing screen.

Maybe it'd be great to reiterate if we can find a more specific name here.

**Comment by KabanFriends** [comment]
Hello, I'm the one who made the original suggestion #9293, and I have another idea for this. How about triggering the event for when player colors the sign text using dyes, inc sac, glow inc sac etc?

File: patches/api/0419-Add-event-for-player-editing-sign.patch [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
++     * @return {@link Sign} that was clicked
++     */
++    @NotNull
++    public Sign getSign() {
++        return sign;
++    }
++
++    /**
++     * Get which side of the sign was clicked
++     * @return {@link Side} which was clicked
```
Comment (Lulu13022002) [inline patches/api/0419-Add-event-for-player-editing-sign.patch]: Missing a newline between the description and the return

File: patches/server/0977-Add-event-for-player-editing-sign.patch [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+
+diff --git a/src/main/java/net/minecraft/world/level/block/SignBlock.java b/src/main/java/net/minecraft/world/level/block/SignBlock.java
+index 91bb294be2cd8ab3467a62006a5a2751e0bec4ba..7909c3404b235c8259f9f9decd198e68e8ab9397 100644
+--- a/src/main/java/net/minecraft/world/level/block/SignBlock.java
++++ b/src/main/java/net/minecraft/world/level/block/SignBlock.java
+@@ -3,6 +3,7 @@ package net.minecraft.world.level.block;
+ import java.util.Arrays;
+ import java.util.UUID;
+ import javax.annotation.Nullable;
++
```
Comment (Lulu13022002) [inline patches/server/0977-Add-event-for-player-editing-sign.patch]: Unnecessary diff

File: patches/server/0977-Add-event-for-player-editing-sign.patch (hunk 4 above)
Comment (by77er) [reply patches/server/0977-Add-event-for-player-editing-sign.patch]: Added a commit that flattens the logic out in a more readable way. LMK if this is what you meant!

Comment (lynxplay) [reply patches/server/0977-Add-event-for-player-editing-sign.patch]: Might be a bit too much imo. The two diffs I posted above yank a bunch but leave the core logic that follows in tact.
Tho I'd like machines input on this too in regards to what he wanted :+1: 

Thanks for the work already tho!

Comment (by77er) [reply patches/server/0977-Add-event-for-player-editing-sign.patch]: Of course. Whatever works best!

**Review by bdotsamir — COMMENTED** [review]
I remember looking at this file when I was going to try my hand at fixing the issue and thinking "damn these variable names are awful." thank you for cleaning them up, but isn't the goal to make the diff as small as possible? would a change like this warrant ignoring that goal?

edit since it didn't appear properly: I'm referring to the SignBlock class

**Comment by by77er** [comment]
@bdotsamir Yeah, minimizing the diff makes sense to me. I think I'll push with @lynxplay's change. I'll hang on to the rewritten version just in case that's the way we decide to go. Still gotta decide on a good name for the event.

File: patches/api/0419-Add-event-for-player-editing-sign.patch (hunk 7 above)
Comment (by77er) [reply patches/api/0419-Add-event-for-player-editing-sign.patch]: Maybe just `PlayerOpenSignEvent`? Least clunky name I could think of, if a little vague.

File: patches/server/0977-Add-event-for-player-editing-sign.patch [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+-        Item signBlockEntity = itemStack.getItem();
+-        SignApplicator var10000;
+-        if (signBlockEntity instanceof SignApplicator signApplicator) {
+-            var10000 = signApplicator;
+-        } else {
+-            var10000 = null;
+-        }
+-
+-        SignApplicator signApplicator2 = var10000;
++        SignApplicator signApplicator2 = item instanceof SignApplicator signApplicator ? signApplicator : null; // Paper - decompile fixes
```
Comment (Lulu13022002) [inline patches/server/0977-Add-event-for-player-editing-sign.patch]: This should go into the right patch (MC-Dev fixes)

File: patches/api/0419-Add-event-for-player-editing-sign.patch (hunk 8 above)
Comment (by77er) [reply patches/api/0419-Add-event-for-player-editing-sign.patch]: Added.

File: patches/server/0977-Add-event-for-player-editing-sign.patch (hunk 9 above)
Comment (by77er) [reply patches/server/0977-Add-event-for-player-editing-sign.patch]: Removed.

File: patches/server/0977-Add-event-for-player-editing-sign.patch (hunk 10 above)
Comment (by77er) [reply patches/server/0977-Add-event-for-player-editing-sign.patch]: Done. Good to know, thanks.

File: patches/server/0975-Add-event-for-player-editing-sign.patch [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+-        player.openTextEdit(blockEntity, front);
++        // Paper start
++        org.bukkit.entity.Player bukkitPlayer = (org.bukkit.entity.Player) player.getBukkitEntity();
++        org.bukkit.block.Block bukkitBlock = org.bukkit.craftbukkit.block.CraftBlock.at(blockEntity.getLevel(), blockEntity.getBlockPos());
++        org.bukkit.craftbukkit.block.CraftSign<?> bukkitSign = (org.bukkit.craftbukkit.block.CraftSign<?>) org.bukkit.craftbukkit.block.CraftBlockStates.getBlockState(bukkitBlock);
++        io.papermc.paper.event.player.PlayerOpenSignEvent event = new io.papermc.paper.event.player.PlayerOpenSignEvent(
++            bukkitPlayer,
++            bukkitSign,
++            front ? org.bukkit.block.sign.Side.FRONT : org.bukkit.block.sign.Side.BACK);
++        if (event.callEvent()) {
```
Comment (lynxplay) [inline patches/server/0975-Add-event-for-player-editing-sign.patch]: Could technically reduce the diff here by changing this to a 
```java
if(!event.callEvent()) return
```

**Review by lynxplay — COMMENTED** [review]
Generally looks good to me, I'll be testing it tomorrow to hopefully give it an approval :)
Thanks already for the great work!

File: patches/server/0975-Add-event-for-player-editing-sign.patch (hunk 11 above)
Comment (by77er) [reply patches/server/0975-Add-event-for-player-editing-sign.patch]: True! Done.

**Comment by molor** [comment]
(do not forget this PR please, we're waiting very much for it.. :c also thanks for all your work!)

**Review by kennytv — APPROVED** [review]

**Comment by by77er** [comment]
Rebased and ready.

**Review by electronicboy — APPROVED** [review]

**Comment by ammodev** [comment]
Using signSide.line(component) has no effect on the resulting gui, does it?
